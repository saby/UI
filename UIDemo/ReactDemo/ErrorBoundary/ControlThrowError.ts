import {Control} from 'UICore/Base';
import * as React from 'react';

// tslint:disable:ban-ts-ignore
// tslint:disable:no-any
export default class ControlThrowError extends Control {
    constructor(props: any, ctx: any) {
        super(props, ctx);
        // @ts-ignore
        this.state = { shouldThrowError: false };
        this.throwError = this.throwError.bind(this);
    }


    throwError(): void {
        // @ts-ignore
        this.setState(() => ({
            shouldThrowError: true
        }));
    }
    render(): any {
        // @ts-ignore
        if(this.state.shouldThrowError){
            throw new Error('i crashed');
        }
        return React.createElement('div',{onClick: this.throwError}, 'click to throw error');
    }
}
