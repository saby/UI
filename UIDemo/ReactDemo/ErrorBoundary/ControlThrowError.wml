<div class="control-throw-error" on:click="increaseCounter()">
    some
</div>
