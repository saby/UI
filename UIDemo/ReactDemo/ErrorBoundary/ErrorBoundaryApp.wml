<div class='errorboundary-app-tmplte'>
    <UIDemo.ReactDemo.ErrorBoundary.ControlThrowError />
</div>


