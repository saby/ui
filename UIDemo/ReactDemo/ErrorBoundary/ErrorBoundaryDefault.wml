<div>
    <span>
    Использовать обработку ошибки по дефолту:
    </span>
    <UIDemo.ReactDemo.ErrorBoundary.ControlThrowError/>
</div>
