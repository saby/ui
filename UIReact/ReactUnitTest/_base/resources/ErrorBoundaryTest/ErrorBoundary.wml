<div>
    <ReactUnitTest._base.resources.ErrorBoundaryTest.CrashedControl/>
</div>
