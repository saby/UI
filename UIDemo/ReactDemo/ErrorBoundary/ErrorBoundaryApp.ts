import { Control } from 'UICore/Base';
import { TemplateFunction } from 'UI/Base';
import * as template from "wml!UIDemo/ReactDemo/ErrorBoundary/ErrorBoundaryApp";

export default class ErrorBoundaryApp extends Control {
    protected _template: TemplateFunction = template;
}
