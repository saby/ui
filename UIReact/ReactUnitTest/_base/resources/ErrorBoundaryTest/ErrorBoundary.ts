import { Control, TemplateFunction } from 'UI/Base';
import * as template from 'wml!ReactUnitTest/_base/resources/ErrorBoundaryTest/ErrorBoundary';

export default class ErrorBoundary extends Control {
    _template: TemplateFunction = template;
}
