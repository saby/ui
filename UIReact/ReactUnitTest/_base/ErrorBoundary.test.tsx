/**
 * @jest-environment jsdom
 */
import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import ErrorBoundary from './resources/ErrorBoundaryTest/ErrorBoundary';
import {
    functionNameForTestStack,
    errorMessage,
} from './resources/ErrorBoundaryTest/CrashedControl';

describe('ErrorBoundary', () => {
    let container;
    let consoleMock: jest.SpyInstance;
    beforeEach(() => {
        // подготавливаем DOM-элемент, куда будем рендерить
        container = document.createElement('div');
        document.body.appendChild(container);
        // мокаем console.error т.к. React таким образом пишет ошибку в render контрола CrashedControl
        consoleMock = jest.spyOn(console, 'error').mockImplementation(() => undefined);
    });
    afterEach(() => {
        // подчищаем после завершения
        unmountComponentAtNode(container);
        container.remove();
        container = null;
        jest.restoreAllMocks();
    });
    it('ErrorBoundary должен отрисовать заглушку', () => {
        act(() => {
            render(<ErrorBoundary />, container);
        });
        expect(container).toMatchSnapshot();
    });
    it('Ошибка в консоль содержит оригинальный стек', () => {
        act(() => {
            render(<ErrorBoundary />, container);
        });
        const testingErrorCall = consoleMock.mock.calls.find(
            (call) => call[0].includes('CONTROL ERROR') && call[1].includes(errorMessage)
        );
        expect(testingErrorCall).toBeTruthy();
        expect(testingErrorCall[2]).toContain(functionNameForTestStack);
    });
});
