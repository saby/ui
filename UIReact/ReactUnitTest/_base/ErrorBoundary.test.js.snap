// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary ErrorBoundary должен отрисовать заглушку 1`] = `
<div>
  <div
    style="padding: 50px; max-width: 30%; margin: 0px auto; text-align: center;"
  >
    <img
      src="/cdn/Maintenance/1.1.0/img/NOT_FOUND_default.svg"
      style="max-width: 100%; max-height: 200px; margin-bottom: 20px;"
    />
    <div>
      Что-то пошло не так
    </div>
  </div>
</div>
`;
